src/control/lib/hardware/hwloc/provider.go

```diff
@@ -194,42 +258,45 @@ func (p *Provider) getPCIDevsPerNUMANode(topo *topology) map[uint]hardware.PCIDe
 
 		osDevType, err := osDev.osDevType()
 		if err != nil {
-			// shouldn't be possible
-			p.log.Error(err.Error())
-			continue
+			return err
 		}
 
+		var addr *common.PCIAddress
+		var linkSpeed float64
 		switch osDevType {
 		case osDevTypeNetwork, osDevTypeOpenFabrics:
-			addr := "none"
 			if pciDev, err := osDev.getAncestorByType(objTypePCIDevice); err == nil {
 				addr, err = pciDev.pciAddr()
 				if err != nil {
-					// shouldn't be possible
-					p.log.Error(err.Error())
+					return err
+				}
+				linkSpeed, err = pciDev.linkSpeed()
+				if err != nil {
+					return err
 				}
 			} else {
 				// unexpected - network devices should be on the PCI bus
 				p.log.Error(err.Error())
-
 			}
+		default:
+			continue
+		}
 
-			numaID := p.getDeviceNUMANodeID(osDev, topo)
-
-			if _, exists := pciPerNUMA[numaID]; !exists {
-				pciPerNUMA[numaID] = make(hardware.PCIDevices)
+		numaID := p.getDeviceNUMANodeID(osDev, topo)
+		for _, node := range nodes {
+			if node.ID != numaID {
+				continue
 			}
-
-			pciPerNUMA[numaID].Add(&hardware.Device{
-				Name:         osDev.name(),
-				Type:         osDevTypeToHardwareDevType(osDevType),
-				PCIAddr:      addr,
-				NUMAAffinity: numaID,
+			node.AddDevice(&hardware.PCIDevice{
+				Name:      osDev.name(),
+				Type:      osDevTypeToHardwareDevType(osDevType),
+				PCIAddr:   *addr,
+				LinkSpeed: linkSpeed,
```

Comment:
would it also be possible to get PCIe link width in a similar manner?

Reply:
Not with hwloc, I don't think. But we can probably add a pciutils provider to get at that info.